### PR TITLE
Added the Ability to Use Rectilinear Polygons in the Dynamic Zone Block

### PR DESCRIPTION
**Pull request by chandlersupple**
# Description
Added two optional parameters to the Dynamic Zone block allowing users to use rectangles instead of irregular polygons, and get the height, width, and angle of the rectangles.

dynamic-zone-block-pr

## Type of change
-   [X] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

Passed all tests. Added the `test_calculate_minimum_bounding_rectangle` test.

## Any specific deployment considerations

N/A

## Docs

N/A

### CODE REVIEW

File: inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py [hunk 1]
```diff
@@ -77,24 +85,42 @@ def get_execution_engine_compatibility(cls) -> Optional[str]:
...
+    mask: np.ndarray,
+    required_number_of_vertices: int,
+    force_rectangle: bool = False,
+    max_steps: int = 1000,
+) -> Union[np.array, Tuple[np.array, float, float, float]]:
     contours = sv.mask_to_polygons(mask)
     largest_contour = max(contours, key=len)
 
-    # https://docs.opencv.org/4.x/d3/dc0/group__imgproc__shape.html#ga014b28e56cb8854c0de4a211cb2be656
+    if force_rectangle and required_number_of_vertices == 4:
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py:112]: I'd move this to the `run` function and keep all logic relevant to forced rectangle together

File: inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py [hunk 2]
```diff
@@ -137,9 +165,24 @@ def run(
                 simplified_polygon = calculate_simplified_polygon(
                     mask=mask,
                     required_number_of_vertices=required_number_of_vertices,
+                    force_rectangle=force_rectangle,
                 )
-                if len(simplified_polygon) != required_number_of_vertices:
-                    continue
-                simplified_polygons.append(simplified_polygon)
+                if force_rectangle and required_number_of_vertices == 4:
```
Comment (grzegorz-roboflow) [inline inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py:170]: I'd vote to raise exception at the beginning of `run` method if `force_rectangle` is true and requested number of vertices was not set to 4.

**Review by grzegorz-roboflow — CHANGES_REQUESTED** [review]

**Comment by PawelPeczek-Roboflow** [comment]
@chandlersupple could you please follow up on @grzegorz-roboflow comments - otherwise it may not be pushed into next release

File: inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py (hunk 2 above)
Comment (chandlersupple) [reply inference/core/workflows/core_steps/transformations/dynamic_zones/v1.py:170]: Got it. Thanks for flagging this @grzegorz-roboflow. I'll move the logic to the `run` function.

**Review by PawelPeczek-Roboflow — DISMISSED** [review]
I am approving this change, but I have one observation regarding interoperability of the block. 

We claim the output is of kind `list_of_values` which is true, but `include_rectangle_details` changes the output data - making downstream blocks compatible or incompatible - depending on the value of this parameter. I am not sure how it ends up working - maybe that would be a problem, maybe not. @chandlersupple could you please monitor that over time and when we see people struggling with adoption of the block due to that inconsistency ping me to solve the problem.

**Comment by grzegorz-roboflow** [comment]
> I am approving this change, but I have one observation regarding interoperability of the block.
> 
> We claim the output is of kind `list_of_values` which is true, but `include_rectangle_details` changes the output data - making downstream blocks compatible or incompatible - depending on the value of this parameter. I am not sure how it ends up working - maybe that would be a problem, maybe not. @chandlersupple could you please monitor that over time and when we see people struggling with adoption of the block due to that inconsistency ping me to solve the problem.

@chandlersupple I moved all logic related to bounding rect into new block

File: inference/core/workflows/core_steps/transformations/bounding_rect/v1.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+This block is best suited when Zone needs to be created based on shape of detected object
+(i.e. basketball field, road segment, zebra crossing etc.)
+Input detections should be filtered beforehand and contain only desired classes of interest.
+"""
+
+
+class BoundingRectManifest(WorkflowBlockManifest):
+    model_config = ConfigDict(
+        json_schema_extra={
+            "name": "Dynamic Zone",
```
Comment (PawelPeczek-Roboflow) [inline inference/core/workflows/core_steps/transformations/bounding_rect/v1.py:38]: name is copy pasted and no not match

**Review by PawelPeczek-Roboflow — COMMENTED** [review]
please add end-to-end test to integration tests and add to gallery

**Review by grzegorz-roboflow — DISMISSED** [review]

**Review by grzegorz-roboflow — APPROVED** [review]

**Comment by hansent** [comment]
I agree this should be a separate block.  Noticed that the output kind changed (was LIST_OF_VALUES via dynamic zone block, but new block has sv.detections as output); maybe that's what we want, but not sure what the desired use case is.  @chandlersupple how / what is going to be using the rectangle data?

Main thing I can think of is difference between predictions and zones (for e.g. time in zone block), maybe we can have a block to convert between these.

**Review by PawelPeczek-Roboflow — APPROVED** [review]

**Pull request merged**